### PR TITLE
FEAT: Introduce integration with Algolia

### LINKED ISSUE #932: Introduce Algolia Integration
**Is your feature request related to a problem? Please describe.**
Many developers, including myself, use [Algolia](https://www.algolia.com/) for search functionalities. Integrating Algolia into Splainer-Search would be beneficial for debugging and optimizing Algolia search queries, expanding the tool's utility.

**Describe the solution you'd like**
Support for Algolia API integration in https://github.com/o19s/splainer-search must be introduced. Quepid must have the ability to configure the endpoint from the UI.

**Describe alternatives you've considered**
I have tried using Custom Search API as an alternative. Algolia supports querying a single index and multiple indices at the same time. The Custom Search API endpoint would require custom mapper functions that would become complex over time. Since Vectara was recently introduced in Quepid, it made sense to introduce native support for Algolia.

**Additional context**

image

I have opened a pull request o19s/splainer-search/pull/145 to introduce support in splainer-search. This issue would be a continuation to introduce the support in Quepid.

### PR DESCRIPTION
**Pull request by sumitsarkar**
## Description
Introducing Algolia Search Endpoint in Quepid. This would allow users to configure Algolia Endpoint to query a single index. 
Multiple indices query is not implemented.
Refer to o19s/splainer-search#144 for the related issue in splainer-search. Once the splainer-search library is released, this PR would be able to add the integration.

## Motivation and Context
Closes #932 

## How Has This Been Tested?
All tests were manually done. I am open to hear from maintainers on what kind of automated tests can be introduced to reduce this overhead.
1. Used a dev version of splainer-search(o19s/splainer-search#145) on Quepid installation.
2. Created a few Algolia endpoints, and ran searches on an index to list out results, and score them.
3. Tried cloning a few Algolia endpoints.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves existing functionality)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by epugh** [comment]
Use of `lambda` is cool!   First time I've seen this!

**Comment by sumitsarkar** [comment]
> Use of `lambda` is cool! First time I've seen this!

I don't know if it's a good way to solve this. This is the first time I am writing Ruby code. If there are better alternatives to this, happy to incorporate them in the PR. 

Btw, the [JS analysis is failing](https://app.deepsource.com/gh/o19s/quepid/run/1a5c3f2b-bab7-4433-aba8-345e05ce7317/javascript/JS-D008?listindex=0). I suspect it's a false positive as that's how Angular modules are defined. Am I missing something?

**Review by epugh — COMMENTED** [review]
Looking great!

**Comment by epugh** [comment]
> > Use of `lambda` is cool! First time I've seen this!
> 
> I don't know if it's a good way to solve this. This is the first time I am writing Ruby code. If there are better alternatives to this, happy to incorporate them in the PR.
> 
> Btw, the [JS analysis is failing](https://app.deepsource.com/gh/o19s/quepid/run/1a5c3f2b-bab7-4433-aba8-345e05ce7317/javascript/JS-D008?listindex=0). I suspect it's a false positive as that's how Angular modules are defined. Am I missing something?

The js tool will "re evaluate" code that gets changed, so the underlying pattern of the failure was already there, and I think we can ignore that one.

File: app/assets/javascripts/services/docCacheSvc.js [hunk 1]
```diff
@@ -58,21 +58,24 @@ angular.module('QuepidApp')
...
-        
+
         // 'vectara' does not support doc lookup by ID.
         let supportLookupById = true;
         if (settings && settings.searchEngine === 'vectara'){
           supportLookupById = false;
         }
         else if (settings && settings.searchEngine === 'searchapi'){
           supportLookupById = false;
+        } else if (settings && settings.searchEngine === 'algolia') {
```
[the remark this answers is not in the document]
Comment (sumitsarkar) [reply app/assets/javascripts/services/docCacheSvc.js:72]: Sure. I have updated the PR for splainer-search o19s/splainer-search/145. 

I have pushed an update for changing this flag as well.

**Comment by epugh** [comment]
Thanks @sumitsarkar  for your work, I put a thank you in https://github.com/o19s/quepid/wiki/Extending-Quepid-for-Your-Search-Engine

**Pull request merged**